flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -81,16 +92,12 @@ NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connection
 				Object old = clients.putIfAbsent(connectionId, connectingChannel);
 
 				if (old == null) {
-					nettyClient.connect(connectionId.getAddress()).addListener(connectingChannel);
-
-					client = connectingChannel.waitForChannel();
-
-					clients.replace(connectionId, connectingChannel, client);
+					connectChannelWithRetry(connectingChannel, connectionId);
+					client = (NettyPartitionRequestClient) clients.get(connectionId);
 				}
 				else if (old instanceof ConnectingChannel) {
-					client = ((ConnectingChannel) old).waitForChannel();
-
-					clients.replace(connectionId, old, client);
+					connectChannelWithRetry((ConnectingChannel) old, connectionId);
```

Comment:
I have some concerns with this operation. `nettyClient.connect(connectionId.getAddress())` should be called only once if multiple threads invoking the `createPartitionRequestClient` at the same time. But now the `#connect` might be called multiple times by different threads to establish multiple physical connections.

E.g. when the first one `clients.putIfAbsent(connectionId, connectingChannel)` to enter `if (old == null)` path, then it would call `connectChannelWithRetry` to establish a connection.

The second one `clients.putIfAbsent(connectionId, connectingChannel)` to get `old instanceof ConnectingChannel` to enter this path, then it also calls `connectChannelWithRetry` to establish another connection.

Reply:
Good catch. It is a bug. There should be a flag in parameter to check whether to connect or not.